sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@intel.com>
+//
+
+#include <sound/soc.h>
+#include <sound/sof/ipc4/header.h>
+#include <uapi/sound/sof/header.h>
+#include "sof-priv.h"
+#include "ipc4-priv.h"
+#include "sof-client.h"
+#include "sof-client-probes.h"
+#include "ipc4-topology.h"
+
+enum sof_ipc4_dma_type {
+	SOF_IPC4_DMA_HDA_HOST_OUTPUT = 0,
+	SOF_IPC4_DMA_HDA_HOST_INPUT = 1,
+	SOF_IPC4_DMA_HDA_LINK_OUTPUT = 8,
+	SOF_IPC4_DMA_HDA_LINK_INPUT = 9,
+	SOF_IPC4_DMA_DMIC_LINK_INPUT = 11,
+	SOF_IPC4_DMA_I2S_LINK_OUTPUT = 12,
+	SOF_IPC4_DMA_I2S_LINK_INPUT = 13,
+};
+
+union sof_ipc4_connector_node_id {
+	u32 val;
+	struct {
+		u32 vindex:8;
+		u32 dma_type:5;
+		u32 rsvd:19;
+	};
+} __packed __aligned(4);
+
+enum sof_ipc4_probe_runtime_param {
+	SOF_IPC4_PROBE_INJECTION_DMA = 1,
+	SOF_IPC4_PROBE_INJECTION_DMA_DETACH,
+	SOF_IPC4_PROBE_POINTS,
+	SOF_IPC4_PROBE_POINTS_DISCONNECT,
+};
+
+/* For getting probe dmas */
+struct sof_ipc4_probe_dma {
+	union sof_ipc4_connector_node_id node_id;
+	unsigned int dma_buffer_size;
+} __packed __aligned(4);
+
+struct sof_ipc4_probe_gtw_cfg {
+	union sof_ipc4_connector_node_id node_id;
+	u32 dma_buffer_size;
+} __packed __aligned(4);
+
+struct sof_ipc4_probe_cfg {
+	struct sof_ipc4_base_module_cfg base;
+	struct sof_ipc4_probe_gtw_cfg gtw_cfg;
+} __packed __aligned(4);
+
+enum sof_ipc4_probe_type {
+	SOF_IPC4_PROBE_TYPE_INPUT = 0,
+	SOF_IPC4_PROBE_TYPE_OUTPUT,
+	SOF_IPC4_PROBE_TYPE_INTERNAL
+};
+
+union sof_ipc4_probe_point_id {
+	u32 full_id;
+	struct {
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
+		u32 index:6;
+	} fields;
+} __packed __aligned(4);
+
+struct sof_ipc4_probe_point {
+	union sof_ipc4_probe_point_id buffer_id;
+	u32 purpose;
```

Comment:
It depends whether we are in IPC3 [1] or IPC4 [2] mode 😅. In any case the current Linux driver probes implementation tries to be agnostic about the FW details and expose bare numbers though debugfs. The debugfs is then in turn used by user space tools compiled from sof FW sources, sharing the same headers. This is after all debug interface for developers.

Hmm, maybe we should start purpose counting from 0 on IPC3 side too, since the IPC3 API has hardly been ever used outside SOF team.

[1] https://github.com/thesofproject/sof/blob/1bca97ad4cf4e8eb45bd4ce6ca3ac87b499914ee/src/include/ipc3/probe.h#L29

[2] https://github.com/thesofproject/sof/blob/1bca97ad4cf4e8eb45bd4ce6ca3ac87b499914ee/src/include/ipc4/probe.h#L38

Reply:
This is IPC4 probes code, so what IPC3 defines is irrelevant? If you are keeping the debugfs interface generic, then either IPC3 or IPC4 must align with the core's convention.

Reply:
@ujfalusi @plbossart The probe points are uint32 identifiers. Their allocation policy is different in IPC3 and IPC4, but the current debugfs interface is agnostic to this. In both IPC3 and IPC4, user of debugfs has to have some out-of-band means to figure out which IDs are correct, so no change here because of IPC4.

I do think we need to consider a more user-friendly debugfs interface in the future, but I think we can do that on top. E.g. add introspection (you can query which probe points are available in the system) and then a more user-friendly (and less error prone way) to set them. But I'd say this is beyond this PR (as situation was not much better with IPC3).